### ISSUE TITLE
[CAL-1928] /event-types page slow

### ISSUE THREAD

**Issue by PeerRich** [calcom/cal.com#9448]
find ways to improve event types endpoint and page

when opening a PR make sure to have before and after benchmarks

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1928](https://linear.app/calcom/issue/CAL-1928/event-types-page-slow)</sub>

**Comment by rjackson** [comment]
👀 /attempt #9448

**Comment by piyushgarg-dev** [comment]
@PeerRich, I guess `trpc` is the culprit here 👀

**Comment by rjackson** [comment]
I've been looking into measuring current baseline performance, and trying to set my dev environment up with enough test data to possibly surface any scaling-related bottlenecks.

I'm still working on that, but there are a few things I've spotted that I want to write down as they could be useful to guide our investigations on this:

- Data requests to have a baseline latency of ~280ms for me (delay between me + edge, and edge + db?)
- `/api/trpc/public/i18n` seems generally above baseline, and highly variable
- `/api/trpc/eventTypes/getByViewer` seems to consistently 2-3x above baseline
  - Potentially indicative of database query inefficiencies?
  - This difference is not present on my dev environment (hopefully my test data might surface it later)
  - This endpoint also appears to be re-used across the website, which seems unnecessary. Might be `kbar` functionality?
- `api/trpc/viewer/appById,bookingUnconfirmedCount`appears to be a chained request, initiated after the first chunk of requests
- `/api/trpc/teams/listInvites` appears to be a chained request, initiated after the first chunk of requests

The `public/i18n` and `eventTypes/getByViewer` look like they're worth investigating further, and will be the focus of my attention.

The last 2 endpoints probably aren't a big deal as I'm guessing they will be non-blocking, and relate to smaller UI components.

@PeerRich I don't know if you'd be comfortable sharing how many records each database table has in your prod db?

The following SQL gives counts for each table `eventTypes/getByViewer` loads data from:

```sql
SELECT
  schemaname,
  relname,
  n_live_tup 
FROM pg_stat_user_tables
WHERE relname IN (
  '_user_eventtype',
  'Availability',
  'Credential',
  'DestinationCalendar',
  'EventType',
  'HashedLink',
  'Host',
  'Membership',
  'Schedule',
  'SelectedCalendar',
  'users'
) 
ORDER BY n_live_tup DESC
```

**Comment by sohamdata** [comment]
@rjackson how are you measuring performance?

**Comment by rjackson** [comment]
@sohamdata For the above observations, response times in devtools' network tab.

Locally I'm digging into the execution time of individual database queries.

**Comment by rjackson** [comment]
Darn, I thought I found something but it ended up being a false lead: Loading the `children` relationship on `EventType` executes as a sequential scan with a small amount of `EventType` rows, but as we add more rows it swaps over to using the `[userId, parentId]` index and becomes quick.

Overall, I can't spot anything majorly wrong in the code or in our SQL queries.

The next places I'd advise looking, but unable to do so myself:

- Metrics on how the database itself is performing; make sure its handling its load okay and doesn't need scaling or tweaking
- Log entries for `[PERF]: getByViewer`, to see what the measured performance is for `eventTypes/getByViewer` from the server's perspective.
  - If that's slow, we might be able to narrow the investigation with some more in-depth query logging – I've committed the increased logging I've been using here: ddca97da0bbfec6a160c5a0595157a6505c2bfd7
